astrobase/periodbase/tls.py

```diff
@@ -0,0 +1,380 @@
+# -*- coding: utf-8 -*-
+# tls.py - Luke Bouma (luke@astro.princeton.edu) - Apr 2019
+
+"""
+Contains the Hippke & Heller (2019) transit-least-squared period-search
+algorithm implementation for periodbase. This depends on the external package
+written by Hippke & Heller, https://github.com/hippke/tls.
+"""
+
+#############
+## LOGGING ##
+#############
+
+import logging
+from astrobase import log_sub, log_fmt, log_date_fmt
+
+DEBUG = False
+if DEBUG:
+    level = logging.DEBUG
+else:
+    level = logging.INFO
+LOGGER = logging.getLogger(__name__)
+logging.basicConfig(
+    level=level,
+    style=log_sub,
+    format=log_fmt,
+    datefmt=log_date_fmt,
+)
+
+LOGDEBUG = LOGGER.debug
+LOGINFO = LOGGER.info
+LOGWARNING = LOGGER.warning
+LOGERROR = LOGGER.error
+LOGEXCEPTION = LOGGER.exception
+
+
+#############
+## IMPORTS ##
+#############
+
+import numpy as np
+
+from multiprocessing import Pool, cpu_count
+
+from math import fmod
+
+from numpy import (
+    nan as npnan, arange as nparange, array as nparray,
+    isfinite as npisfinite, argmax as npargmax, linspace as nplinspace,
+    ceil as npceil, argsort as npargsort, concatenate as npconcatenate
+)
+
+from astropy import units as u
+
+try:
+    from transitleastsquares import transitleastsquares
+except:
+    errmsg = (
+        'tried importing transitleastsquares and failed.\n'
+        'are you sure you have installed it correctly?\n'
+        'see https://transitleastsquares.readthedocs.io/en/latest/Installation.html'
+    )
+    raise AssertionError(errmsg)
+
+
+###################
+## LOCAL IMPORTS ##
+###################
+
+from ..lcmath import sigclip_magseries
+
+
+############
+## CONFIG ##
+############
+
+NCPUS = cpu_count()
+
+
+#######################
+## UTILITY FUNCTIONS ##
+#######################
+
+def tls_parallel_pfind(times, mags, errs=None,
+                       magsarefluxes=None,
+                       startp=0.1,  # search from 0.1 d to...
+                       endp=None,   # determine automatically from times
+                       tlsoversample=5,
+                       tlsmintransits=3,
+                       periodepsilon=0.1,
+                       nbestpeaks=5,
+                       sigclip=10.0,
+                       verbose=True,
+                       nworkers=None):
+    """
+    Wrapper to Hippke & Heller (2019)'s "transit least squares", which is BLS,
+    but with a slightly better template (and niceties in the implementation).
+
+    A few comments:
+
+    * The time series must be in units of days.
+
+    * The frequency sampling Hippke & Heller (2019) advocate for is cubic in
+    frequencies, instead of linear. Ofir (2014) found that the
+    linear-in-frequency sampling (which is correct for sinusoidal signal
+    detection) isn't optimal for a Keplerian box signal. He gave an equation
+    for "optimal" sampling. `tlsoversample` is the factor by which to
+    oversample over that. The grid can be imported independently via
```

Comment:
Add a `::` at the end of this line to make Sphinx treat the next indented line as code.